### PR TITLE
fix image preview rotation in ComposeActivity

### LINKED ISSUE #827: Tusky composer doesn't honor orientation metadata
My only takes pictures in landscape and stores its orientation in EXIF metadata. While it looks fine in the timeline once it is posted, it isn't respected in the composer (see screenshot)

screenshot_20180901-145041_tusky

[actual toot](https://octodon.social/@Siphonay/100650598134386494)

* * * *
- Tusky Version: 3.0
- Android Version: 8.0
- Android Device: Samsung Galaxy S9
- Mastodon instance (if applicable): any

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by connyduck**
closes #827 

We already had the code in Tusky, just needed to call it from the right place.
For some reason Picasso ignores orientation when used with uris as well, so I used RxJava instead.

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 1]
```diff
@@ -1121,11 +1130,24 @@ private void makeCaptionDialog(QueuedMedia item) {
...
+                    @Override
+                    public void onSubscribe(Disposable d) {}
+
+                    @Override
+                    public void onSuccess(Bitmap bitmap) {
+                        imageView.setImageBitmap(bitmap);
+                    }
+
+                    @Override
+                    public void onError(Throwable e) { }
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: Did you use `SingleObserver` and not consumer to swallow errors? You could still use `BiConsumer` or two `Consumer`s to save a little bit of space but... idk. I just don't like Java's huge declarations but that's not important.

File: app/src/main/java/com/keylesspalace/tusky/util/MediaUtils.java [hunk 2]
```diff
@@ -176,4 +180,81 @@ public static int calculateInSampleSize(BitmapFactory.Options options, int reqWi
...
+            return result;
+        } catch (OutOfMemoryError e) {
+            return null;
+        }
+    }
+
+    public static int getImageOrientation(Uri uri, ContentResolver contentResolver) {
+        InputStream inputStream;
+        try {
+            inputStream = contentResolver.openInputStream(uri);
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/util/MediaUtils.java]: We could rewrite it as try-with-resources but whatever

**Comment by connyduck** [comment]
Yes I am waiting for Picasso 3 release, this 2.71828 Version doesn't have a changelog

**Comment by Siphonay** [comment]
Good job friends, thank you!

File: app/src/main/java/com/keylesspalace/tusky/util/MediaUtils.java [hunk 3]
```diff
@@ -176,4 +180,81 @@ public static int calculateInSampleSize(BitmapFactory.Options options, int reqWi
...
+        }
+        try {
+            Bitmap result = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(),
+                    bitmap.getHeight(), matrix, true);
+            if (!bitmap.sameAs(result)) {
+                bitmap.recycle();
+            }
+            return result;
+        } catch (OutOfMemoryError e) {
+            return null;
```
Comment (igalic) [inline app/src/main/java/com/keylesspalace/tusky/util/MediaUtils.java]: would giving the user an error just cause another OOM at this point?

Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/util/MediaUtils.java]: No, garbage collector frees up the memory reserved for the bitmap

**Comment by connyduck** [comment]
User see a generic error message when an OOM exception occurs. We could show a more specific error message.

**Pull request merged**